libsolidity/interface/StandardCompiler.cpp

```diff
@@ -1504,94 +1501,109 @@ Json StandardCompiler::compileSolidity(StandardCompiler::InputsAndSettings _inpu
+				evmData["gasEstimates"] = compilerStack.gasEstimates(contractName);
+
+			if (compilationSuccess && isArtifactRequested(
+				_inputsAndSettings.outputSelection,
+				file,
+				name,
+				evmObjectComponents("bytecode"),
+				wildcardMatchesExperimental
+			))
+				evmData["bytecode"] = collectEVMObject(
+					_inputsAndSettings.evmVersion,
+					compilerStack.object(contractName),
+					compilerStack.sourceMapping(contractName),
+					compilerStack.generatedSources(contractName),
+					false,
+					[&](std::string const& _element) { return isArtifactRequested(
+						_inputsAndSettings.outputSelection,
+						file,
+						name,
+						"evm.bytecode." + _element,
+						wildcardMatchesExperimental
+					); }
+				);
 
-		if (compilationSuccess && isArtifactRequested(
-			_inputsAndSettings.outputSelection,
-			file,
-			name,
-			evmObjectComponents("deployedBytecode"),
-			wildcardMatchesExperimental
-		))
-			evmData["deployedBytecode"] = collectEVMObject(
-				_inputsAndSettings.evmVersion,
-				compilerStack.runtimeObject(contractName),
-				compilerStack.runtimeSourceMapping(contractName),
-				compilerStack.generatedSources(contractName, true),
-				true,
-				[&](std::string const& _element) { return isArtifactRequested(
-					_inputsAndSettings.outputSelection,
-					file,
-					name,
-					"evm.deployedBytecode." + _element,
-					wildcardMatchesExperimental
-				); }
-			);
+			if (compilationSuccess && isArtifactRequested(
+				_inputsAndSettings.outputSelection,
+				file,
+				name,
+				evmObjectComponents("deployedBytecode"),
+				wildcardMatchesExperimental
+			))
+				evmData["deployedBytecode"] = collectEVMObject(
+					_inputsAndSettings.evmVersion,
+					compilerStack.runtimeObject(contractName),
+					compilerStack.runtimeSourceMapping(contractName),
+					compilerStack.generatedSources(contractName, true),
+					true,
+					[&](std::string const& _element) { return isArtifactRequested(
+						_inputsAndSettings.outputSelection,
+						file,
+						name,
+						"evm.deployedBytecode." + _element,
+						wildcardMatchesExperimental
+					); }
+				);
 
-		if (!evmData.empty())
-			contractData["evm"] = evmData;
+			if (!evmData.empty())
+				contractData["evm"] = evmData;
 
-		if (!contractData.empty())
+			if (!contractData.empty())
+			{
+				if (!contractsOutput.contains(file))
+					contractsOutput[file] = Json::object();
+				contractsOutput[file][name] = contractData;
+			}
+		}
+		catch (UnimplementedFeatureError const& _unimplementedError)
 		{
-			if (!contractsOutput.contains(file))
-				contractsOutput[file] = Json::object();
-			contractsOutput[file][name] = contractData;
+			errors.emplace_back(formatErrorWithException(
+				compilerStack,
+				_unimplementedError,
+				Error::Type::UnimplementedFeatureError,
+				"general",
+				"Unimplemented feature (" + _unimplementedError.lineInfo() + ")"
+			));
 		}
 	}
+	if (errors.size() > 0)
+		output["errors"] = std::move(errors);
```

Comment:
Also note that there's a very similar handler above, covering parsing/analysis/compilation: https://github.com/ethereum/solidity/blob/8a97fa7a1db1ec509221ead6fea6802c684ee887/libsolidity/interface/StandardCompiler.cpp#L1403

It's redundant with the more specific handlers you added inside of `CompilerStack` so I'd remove it. Or rather remove any actual handling from it and have it just rethrow the exception so that it reaches the handler I suggested above and becomes a fatal error.

I don't expect many `UnimplementedFeatureError`s to end up in this handler anyway. All of those that could carry any source location information will be caught by one of the three handlers you added in `CompilerStack`. I think that only unimplemented errors from AST import will end up here. The import does not use error reporter (so cannot report more than one error) and if it fails it's like failed parsing and all cannot result in any valid output so it's already like a fatal error.